File: lib/yahooGame/paIdFormat.ts
/**
 * canonical `paId` の形式（SSOT）。
 *
 * `{gameId}-{inning}-{表|裏}-{paSeqInHalf}`
 *
 * - **paSeqInHalf** … その半回（表/裏）における **打席の通し番号**（1 始まり）。
 *   Yahoo 一球 `score?index=` のキー（Phase10 行の `bat_order`）および
 *   `textPlayByPlay` 行頭の「`N：`」の **N** と一致する。
 * - **打順（1〜9番のスタメン位置）ではない。**
 *   例: `2021038681-4-表-6` = 4回表の **6番目の打席**（1番・桑原のけん制盗塁死）。
 *   外崎の 6番打者 HR は **`4-表-2`**（2番目の打席）に `pitchEvents` がある。
 */

export type ParsedPaId = {
  gameId: string
  inning: number
  half: "表" | "裏"
  /** 半回内打席通し番号（打順ではない） */
  paSeqInHalf: number
  raw: string
}

const PAID_RE = /^(\d+)-(\d+)-(表|裏)-(\d+)$/

export function parsePaId(paId: string): ParsedPaId | null {
  const m = String(paId ?? "").trim().match(PAID_RE)
  if (!m) return null
  return {
    gameId: m[1]!,
    inning: Number(m[2]),
    half: m[3] as "表" | "裏",
    paSeqInHalf: Number(m[4]),
    raw: String(paId).trim(),
  }
}

export function buildPaId(
  gameId: string,
  inning: number,
  half: "表" | "裏",
  paSeqInHalf: number,
): string {
  return `${gameId}-${inning}-${half}-${paSeqInHalf}`
}

/** `score?index=` の 7 桁プレフィックス（例: 4回表6番目 → `0410600`） */
export function paSeqInHalfToScoreIndexPrefix(
  inning: number,
  half: "表" | "裏",
  paSeqInHalf: number,
): string {
  const tb = half === "表" ? "1" : "2"
  return `${String(inning).padStart(2, "0")}${tb}${String(paSeqInHalf).padStart(2, "0")}00`
}

/** 文字列 sort だと 11回が 2回より前に来るため、構造化比較 */
export function comparePaIdChronological(a: string, b: string): number {
  const pa = parsePaId(a)
  const pb = parsePaId(b)
  if (!pa || !pb) return String(a).localeCompare(String(b), "ja")
  if (pa.inning !== pb.inning) return pa.inning - pb.inning
  if (pa.half !== pb.half) return pa.half.localeCompare(pb.half, "ja")
  return pa.paSeqInHalf - pb.paSeqInHalf
}

File: scripts/pa_id_format.py
# -*- coding: utf-8 -*-
"""canonical paId の形式（SSOT）。末尾は打順ではなく半回内打席通し番号（pa_seq_in_half）。"""

from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Literal

PAID_RE = re.compile(r"^(\d+)-(\d+)-(表|裏)-(\d+)$")


@dataclass(frozen=True)
class ParsedPaId:
    game_id: str
    inning: int
    half: Literal["表", "裏"]
    pa_seq_in_half: int  # 打順ではない
    raw: str


def parse_pa_id(pa_id: str) -> ParsedPaId | None:
    m = PAID_RE.match(str(pa_id or "").strip())
    if not m:
        return None
    return ParsedPaId(
        game_id=m.group(1),
        inning=int(m.group(2)),
        half=m.group(3),  # type: ignore[arg-type]
        pa_seq_in_half=int(m.group(4)),
        raw=str(pa_id).strip(),
    )


def build_pa_id(game_id: str, inning: int, half: str, pa_seq_in_half: int) -> str:
    return f"{game_id}-{inning}-{half}-{pa_seq_in_half}"


def pa_seq_in_half_to_score_index_prefix(inning: str | int, top_bottom: str, pa_seq_in_half: str | int) -> str:
    tb = "1" if top_bottom == "表" else "2"
    return f"{int(inning):02d}{tb}{int(pa_seq_in_half):02d}00"
